examples/training/resnet50/main.cpp

```diff
@@ -0,0 +1,276 @@
+    for (unsigned c = 0; c < IMAGE_COLORS; ++c) {
+      auto bgrc = IMAGE_COLORS - 1 - c;
+      for (unsigned h = 0; h < 32; ++h) {
+        for (unsigned w = 0; w < 32; ++w) {
+          // ResNet BGR color space vs CIFAR RGB.
+          auto val = static_cast<float>(static_cast<uint8_t>(dbInput.get()));
+          // Normalize and scale image.
+          val = (val - imagenetNormMean[c]) * scale / imagenetNormStd[c] + bias;
+          for (unsigned sh = 0; sh < expandImage; ++sh) {
+            for (unsigned sw = 0; sw < expandImage; ++sw) {
+              imagesH.at(
+                  {n, bgrc, sh + h * expandImage, sw + w * expandImage}) = val;
+            }
+          }
+        }
+      }
+    }
+  }
+}
+} // namespace
+
+int main(int argc, char **argv) {
+  llvm::cl::ParseCommandLineOptions(argc, argv,
+                                    " ResNet50 Training Example\n\n");
+
+  // We expect the input to be NCHW.
+  llvm::ArrayRef<size_t> allImagesDims = {CIFAR_NUM_IMAGES, IMAGE_COLORS,
+                                          IMAGE_HEIGHT, IMAGE_WIDTH};
+  llvm::ArrayRef<size_t> batchImagesDims = {miniBatchSize, IMAGE_COLORS,
+                                            IMAGE_HEIGHT, IMAGE_WIDTH};
+  llvm::ArrayRef<size_t> initShapeDims = {1, IMAGE_COLORS, IMAGE_HEIGHT,
+                                          IMAGE_WIDTH};
+  llvm::ArrayRef<size_t> allLabelsDims = {CIFAR_NUM_IMAGES, LABELS_SIZE};
+
+  Module mod;
+  PlaceholderBindings bindings;
+  const char inputName[] = "gpu_0/data";
+  TypeRef inputType = mod.uniqueType(ElemKind::FloatTy, initShapeDims);
+  // Load ResNet model.
+  Function *F = mod.createFunction("resnet50");
+  llvm::outs() << "Loading resnet50 model.\n";
+  llvm::Error errPtr = llvm::Error::success();
+
+  // Loader has randomly initialized trainable weights.
+  Caffe2ModelLoader loader(resnet50Path + "/predict_net.pbtxt",
+                           resnet50Path + "/init_net.pb", {inputName},
+                           {inputType}, *F, &errPtr);
+
+  if (errPtr) {
+    llvm::errs() << "Loader failed to load resnet50 model from path: "
+                 << resnet50Path << "\n";
+    return -1;
+  }
+
+  llvm::outs() << "Creating input/output placeholders.\n";
+  // Create the input layer.
+  auto *A = mod.createPlaceholder(ElemKind::FloatTy, batchImagesDims, "input",
+                                  /* isTraining */ false);
+  bindings.allocate(A);
+  // Create the output layer.
+  auto *E =
+      mod.createPlaceholder(ElemKind::FloatTy, {miniBatchSize, LABELS_SIZE},
+                            "expected", /* isTraining */ false);
+  bindings.allocate(E);
+
+  llvm::outs() << "Preparing for training.\n";
+
+  auto &outputs = loader.getOutputVarsMapping();
+  std::unordered_set<std::string> legalOutputNames;
+  for (auto &output : outputs) {
+    legalOutputNames.insert(legalizeName(output.getKey()));
+  }
+
+  prepareFunctionForTraining(F, bindings, legalOutputNames);
```

Comment:
I would prefer if we could hide this inside `glow::differentiate` or `compile`. If we can infer function outputs as per my other comment, I think we can do this. If not, we'll have to live with it.

Reply:
We have a discussion and decision was - have a separate from loader API that user can use to prepare Function for training. Graph can be crated programmatically, in such a case there is no need to call API at all.